programs/exchange_api/src/exchange_processor.rs

```diff
@@ -192,26 +195,64 @@ impl ExchangeProcessor {
         if &id() == keyed_accounts[FROM_ACCOUNT_INDEX].unsigned_key() {
             to_account.tokens[token] += tokens;
         } else {
-            let mut from_account =
-                Self::deserialize_account(&keyed_accounts[FROM_ACCOUNT_INDEX].account.data)?;
-
-            if &from_account.owner != keyed_accounts[OWNER_INDEX].unsigned_key() {
-                error!("Signer does not own from account");
-                Err(InstructionError::GenericError)?
+            let state: ExchangeState =
+                bincode::deserialize(&keyed_accounts[FROM_ACCOUNT_INDEX].account.data)
+                    .map_err(Self::map_to_invalid_arg)?;
+            match state {
+                ExchangeState::Account(mut from_account) => {
+                    println!("tranfer from account");
+                    if &from_account.owner != keyed_accounts[OWNER_INDEX].unsigned_key() {
+                        error!("Signer does not own from account");
+                        Err(InstructionError::GenericError)?
+                    }
+
+                    if from_account.tokens[token] < tokens {
+                        error!("From account balance too low");
```

Comment:
are there tests for these error cases? ideally would have an error code for these scenarios and not do an error!. long term goal maybe once the logic is more finalized.